# NHANES Demographics data

With the goal of sparking discussion surrounding responsible use of machine learning, especially as it applies to health data, it might make sense to first look at what the data says about us, even when it's been anonymized.

Using the NHANES demographic data, is it possible to find small pockets of unique sets of observations that could potentially lead to 'removing the veil' if we we know enough information about someone. This can highlight just how easy it is uniquely describe a person, given enough data

In [40]:
import pandas as pd

In [41]:
data = pd.read_sas('data/P_DEMO.XPT')
data

SEQN  SDDSRVYR  RIDSTATR  RIAGENDR  RIDAGEYR  RIDAGEMN  RIDRETH1  \
0      109263.0      66.0       2.0       1.0       2.0       NaN       5.0   
1      109264.0      66.0       2.0       2.0      13.0       NaN       1.0   
2      109265.0      66.0       2.0       1.0       2.0       NaN       3.0   
3      109266.0      66.0       2.0       2.0      29.0       NaN       5.0   
4      109267.0      66.0       1.0       2.0      21.0       NaN       2.0   
...         ...       ...       ...       ...       ...       ...       ...   
15555  124818.0      66.0       2.0       1.0      40.0       NaN       4.0   
15556  124819.0      66.0       2.0       1.0       2.0       NaN       4.0   
15557  124820.0      66.0       2.0       2.0       7.0       NaN       3.0   
15558  124821.0      66.0       2.0       1.0      63.0       NaN       4.0   
15559  124822.0      66.0       2.0       1.0      74.0       NaN       2.0   

       RIDRETH3  RIDEXMON  DMDBORN4  ...  FIAINTRP  MIALANG  MIAPROXY  \
0           6.0       2.0       1.0  ...       2.0      NaN       NaN   
1           1.0       2.0       1.0  ...       2.0      1.0       2.0   
2           3.0       2.0       1.0  ...       2.0      NaN       NaN   
3           6.0       2.0       2.0  ...       2.0      1.0       2.0   
4           2.0       NaN       2.0  ...       2.0      NaN       NaN   
...         ...       ...       ...  ...       ...      ...       ...   
15555       4.0       1.0       1.0  ...       2.0      1.0       2.0   
15556       4.0       2.0       1.0  ...       2.0      NaN       NaN   
15557       3.0       2.0       1.0  ...       2.0      NaN       NaN   
15558       4.0       1.0       1.0  ...       2.0      1.0       2.0   
15559       2.0       2.0       2.0  ...       NaN      2.0       2.0   

       MIAINTRP  AIALANGA      WTINTPRP      WTMECPRP  SDMVPSU  SDMVSTRA  \
0           NaN       NaN   7891.762435  8.951816e+03      3.0     156.0   
1           2.0       1.0  11689.747264  1.227116e+04      1.0     155.0   
2           NaN       NaN  16273.825939  1.665876e+04      1.0     157.0   
3           2.0       1.0   7825.646112  8.154968e+03      2.0     168.0   
4           NaN       NaN  26379.991724  5.397605e-79      1.0     156.0   
...         ...       ...           ...           ...      ...       ...   
15555       2.0       1.0  21586.596728  2.166689e+04      1.0     166.0   
15556       NaN       NaN   1664.919253  1.838170e+03      2.0     171.0   
15557       NaN       NaN  14819.783161  1.649781e+04      1.0     157.0   
15558       2.0       1.0   4666.817952  4.853430e+03      1.0     158.0   
15559       2.0       NaN  14024.476634  1.519580e+04      1.0     149.0   

       INDFMPIR  
0          4.66  
1          0.83  
2          3.06  
3          5.00  
4          5.00  
...         ...  
15555      3.82  
15556      0.07  
15557      1.22  
15558      3.71  
15559       NaN  

[15560 rows x 29 columns]

In [42]:
# Remove administrative data, statistical weightings to isolate demographic answers
data = data.drop(['SEQN', 
           'SDDSRVYR',
           'WTINTPRP',
           'WTMECPRP',
           'SDMVPSU',
           'SDMVSTRA',
           ],
         axis=1)
data

RIDSTATR  RIAGENDR  RIDAGEYR  RIDAGEMN  RIDRETH1  RIDRETH3  RIDEXMON  \
0           2.0       1.0       2.0       NaN       5.0       6.0       2.0   
1           2.0       2.0      13.0       NaN       1.0       1.0       2.0   
2           2.0       1.0       2.0       NaN       3.0       3.0       2.0   
3           2.0       2.0      29.0       NaN       5.0       6.0       2.0   
4           1.0       2.0      21.0       NaN       2.0       2.0       NaN   
...         ...       ...       ...       ...       ...       ...       ...   
15555       2.0       1.0      40.0       NaN       4.0       4.0       1.0   
15556       2.0       1.0       2.0       NaN       4.0       4.0       2.0   
15557       2.0       2.0       7.0       NaN       3.0       3.0       2.0   
15558       2.0       1.0      63.0       NaN       4.0       4.0       1.0   
15559       2.0       1.0      74.0       NaN       2.0       2.0       2.0   

       DMDBORN4  DMDYRUSZ  DMDEDUC2  ...  SIAPROXY  SIAINTRP  FIALANG  \
0           1.0       NaN       NaN  ...       1.0       2.0      1.0   
1           1.0       NaN       NaN  ...       1.0       2.0      1.0   
2           1.0       NaN       NaN  ...       1.0       2.0      1.0   
3           2.0       2.0       5.0  ...       2.0       2.0      1.0   
4           2.0       3.0       4.0  ...       2.0       2.0      1.0   
...         ...       ...       ...  ...       ...       ...      ...   
15555       1.0       NaN       5.0  ...       2.0       2.0      1.0   
15556       1.0       NaN       NaN  ...       1.0       2.0      1.0   
15557       1.0       NaN       NaN  ...       1.0       2.0      1.0   
15558       1.0       NaN       2.0  ...       2.0       2.0      1.0   
15559       2.0       4.0       3.0  ...       2.0       2.0      NaN   

       FIAPROXY  FIAINTRP  MIALANG  MIAPROXY  MIAINTRP  AIALANGA  INDFMPIR  
0           2.0       2.0      NaN       NaN       NaN       NaN      4.66  
1           2.0       2.0      1.0       2.0       2.0       1.0      0.83  
2           2.0       2.0      NaN       NaN       NaN       NaN      3.06  
3           2.0       2.0      1.0       2.0       2.0       1.0      5.00  
4           2.0       2.0      NaN       NaN       NaN       NaN      5.00  
...         ...       ...      ...       ...       ...       ...       ...  
15555       2.0       2.0      1.0       2.0       2.0       1.0      3.82  
15556       2.0       2.0      NaN       NaN       NaN       NaN      0.07  
15557       2.0       2.0      NaN       NaN       NaN       NaN      1.22  
15558       2.0       2.0      1.0       2.0       2.0       1.0      3.71  
15559       NaN       NaN      2.0       2.0       2.0       NaN       NaN  

[15560 rows x 23 columns]

In [43]:
# Count unique rows in dataset
uniques = len(data[list(data.columns)].drop_duplicates())
print(uniques)
print(uniques/len(data))

14748
0.947814910025707


In the demographic dataset above, of the 15,560 participants interviewed, 14,748 responses (~95%) are completely unique. Put another way, if you had a complete knowledge of a participant's answers to these questions, you could almost certainly pick them out of this dataset without concern of selecting the wrong person.

For exactly that reason, any information that directly links an observation to an actual person has been stripped from this dataset (and would be a cardinal sin to leave in *any* health dataset). That being said, these questions are fairly high-level demographic questions, and on the surface don't seem that they could be descriptive enough.

Let's look at another dataset and see if we can do the same analysis: